storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java

```diff
@@ -293,11 +293,11 @@ public ValidationResult assignOffsetsNonCompressed(LongRef offsetCounter,
 
         if (timestampType == TimestampType.LOG_APPEND_TIME) {
             maxTimestamp = now;
-            offsetOfMaxTimestamp = initialOffset;
-        }
-
-        if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
-            offsetOfMaxTimestamp = offsetCounter.value - 1;
+            if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
+                offsetOfMaxTimestamp = offsetCounter.value - 1;
```

Comment:
Hmm, we should be consistent among different magic values, right? It seems the semantic for MAX_TIMESTAMP is the first offset with the max timestamp.

Reply:
thanks for the comment!
this is the same behaviour before the previous change affected the results. It differentiated the behaviour based on the magic value. I think the reason is because


> /**
> A record batch is a container for records. In old versions of the record format (versions 0 and 1),
> a batch consisted always of a single record if no compression was enabled, but could contain
> many records otherwise. Newer versions (magic versions 2 and above) will generally contain many records
> regardless of compression.
>*/
> public interface RecordBatch extends Iterable<Record>

So version before magic_value=2 doesn't matter.

Reply:
@johnnychhsu : Yes, that was the original code. But the issue is that that code was no longer correct when [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp) was added. Without KIP-734, it's ok to set offsetOfMaxTimestamp to the last offset for magic V2. However, with KIP-734, it's important to set offsetOfMaxTimestamp to the first offset for both magic V1 and V2. It seems that we missed that part when adding KIP-734. So, I think the logic should be if `timestampType == TimestampType.LOG_APPEND_TIME`, we always set `offsetOfMaxTimestamp` to `initialOffset`.

Reply:
It seems KIP-734 does not define the case we are discussing - return the initial/last offset when there are multi-records having same "max timestamp".  

So how we deal with it? Is it a good practice to discuss compatibility? Or we deal with it as "bug fix" and we add enough document to explain the "correct behavior" ?

I prefer the consistent behavior. Hence, it seems to me that adding document to `OffsetSpec.MaxTimestampSpec` is enough here. The use cases which depends on previous behavior (last offset) should be re-written for this bug fix ( and behavior change).

Reply:
> I prefer the consistent behavior. Hence, it seems to me that adding document to OffsetSpec.MaxTimestampSpec is enough here. The use cases which depends on previous behavior (last offset) should be re-written for this bug fix ( and behavior change).

+1 from me.

Reply:
> It seems KIP-734 does not define the case we are discussing - return the initial/last offset when there are multi-records having same "max timestamp".

Yes, I agree that we need to be consistent. Using the initial offset makes the most intuitive sense to me. We could update KIP-734 with this clarification and also send it to the original vote thread to make sure there is no objection.